Enhance Arriba Security

ViniciusSouza opened the pull request:
- Normalize Identity for Arriba Authorization
- Use Authorization Policy with Claims to protect the Controller

ericmaino committed:
PACKAGE: Add missing dependency add build

ericmaino committed:
FIX: Stop using DateTime and move to DateTimeOffset

This change only targets the Item provider at th present moment, howver a bigger change is likely needed to eradicate the usage of DateTime in the codebase

ericmaino committed:
REFACTOR: List<T> dependency to IList<T>

ericmaino committed:
FIX: Update IItemProvider to be an async implemenation

ericmaino committed:
TEST: Remove outdated Arriba.testsettings

ericmaino committed:
FIX: Possible infinite loop?

ericmaino committed:
PLATFORM: Updating most projects to .NET Core

ericmaino committed:
FIX: Arriba.Range conflict System.Range

ericmaino committed:
BREAKING: Dependency on Microsoft.Win32.Registry

ericmaino committed:
BREAKING: Dependency on WindowsIdentity

ericmaino committed:
DUPLICATION: TfsWorkItemCrawler to WorkItemCrawler

- Duplicate the existing project
- Next steps
   - Remove hard TFS dependencies
   - Migrate to .NET Core

ericmaino committed:
PLATFORM: Remove non-Core supported features

Commit:
BUILD: Adding dotnetcore Github Build Action

ericmaino committed:
REFACTOR: Promote DictionaryValueBag

ericmaino committed:
REFACTOR: Introduce ArribaHttpContextRequest

- Add support for both Owin and HttpContext in order to enable aspnet
  controller support moving forward
- Note this implementation is not fully flushed out as things like
  an identity will need to be added once OAuth support exists

ericmaino committed:
FEATURE: Add support aspnet controllers

ericmaino committed:
FIX: Disable early disposal of stream

- Note this appears to be causin and IO issue when running under kestrel

ericmaino committed:
DEBUG: Update IIS to run on 42784

Commit:
BUILD: Address warnings and other .NET Core build issues

Commit:
TEST: Ignore failing tests. Enable build/test gate

ericmaino committed:
FEATURE: Naive Core Azure DevOps Crawler

ericmaino committed:
REFACTOR: Move to src to better support docker builds

meeteric-mob committed:
FEATURE: Add docker support for local debugging

meeteric-mob committed:
REFACTOR: Extract Arriba.Configuration

ViniciusSouza committed:
FIX: Adding script to execute crawler in infinite loop

If the process exits the Azure Web App using compose fails

ViniciusSouza committed:
FEATURE: Adding Enable Authentication flag

ericmaino committed:
DEBUG: Adding support for appsettings.local.json

ericmaino committed:
FIX: Add support for Tags as a string set

ericmaino committed:
PACKAGE: Updating Newtonsoft.json v9.0.1 to v10.0.1

ericmaino committed:
FEATURE: Adding OAuth authentication support

ericmaino committed:
FEATURE: Adding CORS support

ericmaino committed:
FEATURE: Unify configuration processing

ericmaino committed:
DEBUG: Improve debugging messages on .NET Core

ericmaino committed:
FIX: Path related bugs when running on linux

ViniciusSouza committed:
FIX: Adding Authorization header to XHR method

ViniciusSouza committed:
FIX: Using the frontend url as the service base

When using proxy as NGINX the /data can be used

ViniciusSouza committed:
FIX: Adding mime.types for NGINX

SVG was not properly rendered

ViniciusSouza committed:
FEATURE: Adding docker-compose setup document

ViniciusSouza committed:
FIX: Adding server configuration to .docker\Server

Following the pattern of the other services

ericmaino committed:
FIX: Do not delete existing settings

- The image should honor what is on disk and should not work around accidental
  configuration isseus by deleting them

ericmaino committed:
CONFIG: Use default for the default config

- Removes reference from cseng in the docker file and disk

ericmaino committed:
FIX: Ensure environment variables are honored for additional sources

ericmaino committed:
FIX: Increase column suggestion to show top 25

ericmaino committed:
FIX: Deny create access until authorization bug is fixed

ViniciusSouza committed:
TESTING: Add tests to validate the new service

ViniciusSouza committed:
FIX: Doesn't throw exception when key is not found

ViniciusSouza committed:
FIX: Removed unused variable

ViniciusSouza committed:
FIX: GetTableInformationForUser checks for Reader scope

Added a condition to verify if the user has at least reader scope

ViniciusSouza committed:
FEATURE: Expanding IArribaManagementService

ViniciusSouza committed:
REFACTOR: Implementing IArribaManagementService

ViniciusSouza committed:
FEATURE: Add guard type ParamChecker

ViniciusSouza committed:
REFACTOR: Remove handle chain from ArribaTableRoutesApplication

As the service already check permission and validate the request
I removed the handle chain methods from the routes specified on ArribaManagement

On branch visouza/wip-server-refactor
Changes to be committed:
	modified:   src/Arriba/Arriba.Communication/Server/Application/ArribaManagement.cs

ViniciusSouza committed:
FEATURE: Add type ArribaManagementService

The go of this change is to have a separate type for the service
other than use the ArribaManagement
that was used before to avoid code duplication as the refactor was being done.

ViniciusSouza committed:
REFACTOR: Decouple Authorization from ArribaApplication

ViniciusSouza committed:
REFACTOR: Support for Composition

ArribaManagement uses System.Composition need to perform some
changes to allow the ArribaManagementService usage

ViniciusSouza committed:
REFACTOR: Rename ArribaManagement -> ArribaTableRoutesApplication

ViniciusSouza committed:
REFACTOR: Decouple ArribaManagement from ArribaTableRoutesApplication

ericmaino committed:
Merge pull request #11 from ViniciusSouza/visouza/main-ArribaManagementRefactor-clean-history

REFACTOR: Arriba Management

ericmaino committed:
CONFIG: Clean up docker commands and config

- Removes some unnecessary commands
- Ensures that default config values are not valid to ensure proper overriding happens

ViniciusSouza committed:
PIPELINE: Update linux version to ubuntu-20.04

ViniciusSouza committed:
TESTING: Enabling disabled tests

ViniciusSouza committed:
FIX: Unit test Value_Basic

The casting to double in dotnet core has more precision (three more digits
was added).

The same happened for others casting what make the test failing when
converting from .Net Standard to dotnet core

ViniciusSouza committed:
FIX: Unit test ParseQuotedNewLine

The test was using a Windows new line (carriage return) code '/r/n'
to digest and compare the values.

Instead of using the carriage return code I included the usage of
Environment.NewLine that returns the correct value depending on the
OS (Unix Based/Windows Based).

ericmaino committed:
DOCKER: Update to default to published images

ViniciusSouza committed:
FIX: Unit test DistributionQuery_Rounding

Specifying the expected datetime format used for comparison
Not getting impacted by the system cultural definition.

ericmaino committed:
FEATURE: Enable disbling crawler for base image

ericmaino committed:
FIX: Add offline_access to the default scopes

ericmaino committed:
Merge pull request #14 from ViniciusSouza/visouza/main-tests-fix

TESTING: Enable disabled tests and fix them

ViniciusSouza committed:
PROJECT: Change default namespace to Arriba

ViniciusSouza committed:
FEATURE: Disabling authentication for developer loop

ViniciusSouza committed:
REFACTOR: Remove unused property

ViniciusSouza committed:
REFACTOR: Decouple Start from Program

ericmaino committed:
REFACTOR: Cleanup namspaces on Arriba.WorkItemCrawler

ericmaino committed:
REFACTOR: Promote Arriba.Query into Arriba

ericmaino committed:
TEST: Address test to normalize new line inconsistencies

ericmaino committed:
TEST: Fix a bug where tests fail when in src folder

ericmaino committed:
PROJECT: Adding Arriba.Adapters.Netwonsoft

ViniciusSouza committed:
REFACTOR: Removing unused code

ViniciusSouza committed:
REFACTOR: Start.Configure Use IHostEnvironment

ViniciusSouza committed:
REFACTOR: Move Configuration to Configuration folder

ViniciusSouza committed:
RENAME: ArribaContoller to ArribaController

ViniciusSouza committed:
FEATURE: Add Exception Middleware

ViniciusSouza committed:
FEATURE: Add ArribaResultFilter

ArribaRoutedApplication uses ArribaResponse that envelopes the service result.
This filter will envelop the result for ArribaController

ViniciusSouza committed:
FEATURE: Add new route at nginx for ArribaController

At this point the two implementation will run in parallel.
This way it is possible to check any odd behavior that may occur.

ViniciusSouza committed:
FEATURE: Implement ArribaController

ViniciusSouza committed:
REFACTOR: Use route datav2 at the backend url

Nginx has two routes data and datav2:
**data** uses the ApplicationRoutedTable services
**datav2** users the ArribaController services

The both routes will be running in parallel for some time.

ViniciusSouza committed:
REFACTOR: Using Arriba as namespace

Removing Server from the namespace.

ericmaino committed:
REFACTOR: Move to unified Arriba namespace

ericmaino committed:
REFACTOR: Move all converters into Newtonsoft adapter

ericmaino committed:
REFACTOR: Unifi serialization configuration process

ericmaino committed:
REFACTOR: Move Json reader/writers to adapter assembly

ericmaino committed:
REFACTOR: Remove direct dependency on JsonConvert

ericmaino committed:
REFACTOR: Remove dependency on JsonSerializationException

ericmaino committed:
REFACTOR: Remove direct dependency on StringEnumConverter

ericmaino committed:
TEST: Add test to validate SecureSet serialization

ericmaino committed:
REFACTOR: Remove Newtonsoft JsonArray dependency

ericmaino committed:
REFACTOR: Move SecureSet into separate file

ericmaino committed:
REFACTOR: Renami Arriba.Configuration - Arriba.Adapter.Configuration

ericmaino committed:
REFACTOR: Combine Client & Communication into Composition

- Combining code into a single assembly, most of which will be promoted into Arriba in the near future

ericmaino committed:
REFACTOR: Promote types to Arriba requiring access changes

ericmaino committed:
REFACTOR: Promote types into Arriba project

ericmaino committed:
REFACTOR: Remove direct dependency on MemoryCache

- Introduces IObjectCache and IObjectCacheFactory
- Moves MemoryCache dependency into a very specific implementaiton that can next be moved into an adapter.

ericmaino committed:
REFACTOR: Reducing direct depdendencies on composition libraries

ericmaino committed:
REFACTOR: Promoting classes into Arriba project

ericmaino committed:
TEST: Adding a test to verify healthy composition

ericmaino committed:
REFACTOR: Decouple implementation from composition

ericmaino committed:
PROJECT: Adding Arriba.Adapter.Cache

ericmaino committed:
REFACTOR: Moving MemoryCacheFactory into Arriba.Adapter.Cache

ericmaino committed:
PROJECT: Adding Arriba.Adaapter.Telemetry

ericmaino committed:
REFACTOR: Move DataFlow dependenices into Telemetry adapter

ericmaino committed:
REFACTOR: Push ArribaConvert up into Arriba library

- Note I'm still not a fan of this change, but the plumbing required to address it is more then I'd like take on at the present moment.

ericmaino committed:
REFACTOR: Removing direct Newtonsoft dependency

ericmaino committed:
FIX: Address is issue with composition initialization

ericmaino committed:
PROJECT: New Arriba.Adapter.ApplicationInsights

ericmaino committed:
PACKAGE: Upgrading to Newtonsoft 11.0.2

ViniciusSouza committed:
FEATURE: Check for read permission

ViniciusSouza committed:
REFACTOR: Separate Service from Response calls

ericmaino committed:
FEATURE: Add ApplicationVersion and ServiceName

ViniciusSouza committed:
FEATURE: Translate Exceptions into Responses

ViniciusSouza committed:
REFACTOR: Extract IArribaQueryService

ViniciusSouza committed:
TEST: Refactoring ArribaManagementServiceTests

Creating a ArribaServiceBase to be used as a base class for Management and Query services

ericmaino committed:
FEATURE: Naive App Insights Implementation

ViniciusSouza committed:
REFACTOR: Decouple ArribaQueryServices from ArribaQueryApplication

ViniciusSouza committed:
FEATURE: Use ParamChecker guard methods

ViniciusSouza committed:
FEATURE: Add OAuthExtension

ViniciusSouza committed:
REFACTOR: Use Extension OAuthExtension

ViniciusSouza committed:
REFACTOR: Split out host initalization

ericmaino committed:
FIX: Normalize api endpoints using versioning

ericmaino committed:
Merge pull request #15 from ViniciusSouza/visouza/main-ArribaServerRefactor

FEATURE: Add support for AspNet Controller

ericmaino committed:
REFACTOR: Separate Partition Tests

ericmaino committed:
REFACTOR: Extract PartitionConvert and add tests for Ubuntu issue

ericmaino committed:
BUILD: Remove duplicate build and add common parameters

ericmaino committed:
REFACTOR: Extract VerifyNoColumnsAreMissing

ericmaino committed:
REFACTOR: Rename values : datablock

ericmaino committed:
REFACTOR: Improve Partition sorting readability

ericmaino committed:
REFACTOR: Improve partition insert readability

ericmaino committed:
FIX: Ensure partitions are inserted in order

ericmaino committed:
Merge pull request #17 from ericmaino/ericmai/adapter-refactor

REFACTOR: Structures code to better support ports and adapters pattern

ViniciusSouza committed:
TEST: Unit test for ArribaQueryServices

ericmaino committed:
Merge pull request #18 FEATURE: Enable Basic Application Insights

FEATURE: Enable Basic Application Insights

ericmaino committed:
PROJECT: Removing old unused solutions

ViniciusSouza committed:
REFACTOR: Consolidating contract IArribaAuthorization

ViniciusSouza committed:
REFACTOR: Expand contract IArribaServerConfiguration

ViniciusSouza committed:
FIX: Use IArribaServerConfiguration

ViniciusSouza committed:
FIX: Remove unused property _claimsAuth

ViniciusSouza committed:
REFACTOR: Rename namespace to Arriba.Server.Authorization

ericmaino committed:
PACKAGE: Adding DependencyInjection

ViniciusSouza committed:
FIX: Initializing container at constructor

Without having the composition container initialized it is not possible
to add types to it

ViniciusSouza committed:
FIX: Throwing exception when in Developer mode

ViniciusSouza committed:
REFACTOR: Extract security properties to ISecurityConfiguration

ViniciusSouza committed:
FIX: Use one instance of composition host

ViniciusSouza committed:
FEATURE: Add ArribaAuthorizationGrantDecorator

Overrides the returned permission when Authentication is not enabled.

ericmaino committed:
REFACTOR: Unify on the AspNET Composition Model

ericmaino committed:
PACKAGE: Remove Composition Dependencies

ericmaino committed:
Merge pull request #21 from authorization-decorator

Merge pull request #21 from ViniciusSouza/visouza/authorization-refactor

ericmaino committed:
REFACTOR: Composition to use explcit defiintions

ericmaino committed:
Merge pull request #22 REFACTOR: Unify on ASP.NET Composition Model

REFACTOR: Unify on ASP.NET Composition Model

ViniciusSouza committed:
FIX: Initializing ArribaServices

ViniciusSouza committed:
FIX: Using Newtonsoft.Json as the default serializer

By default the AspNet.Core Serializer on 3.0+ is the System.Text.Json, the current implementation of the library doesn't support Fields only Properties even public fields.

There is a change that will be part of .net 5.0 that will provide a option to change this behaviour
https://github.com/dotnet/runtime/pull/36986

ViniciusSouza committed:
REFACTOR: Separate Monitor contract to ITelemetry

ViniciusSouza committed:
FEATURE: Add ArribaQueryController

ViniciusSouza committed:
FEATURE: Support for ArribaRoutes V1

Some Rest API routes was not well formed, in order to move forward to V2 without breaking the UI
was added a ArribaActionController that demanded a Refactor on Arriba's controllers

ericmaino committed:
Merge pull request #26 Enable Quere through ArribaController

Add support for Asp.Net controller for Query

Hugos-MSFT committed:
REFACTOR: Console and trace writeline with ArribaLog.WriteLine.

ViniciusSouza committed:
REFACTOR: OAuthExtension using IAuthorizationHandler

To allow the usage of Authorization attribute at the Controller/Method definition
It will provide a better way to the developer to make changes

mattklup commented on src/Arriba/Arriba.Composition/Composition/Host.cs:
```diff
@@ -60,7 +61,19 @@ private static void AddSecureDatabase(this IServiceCollection services)
+                security.Grant(IdentityScope.Group, PermissionScope.Reader.ToString(), PermissionScope.Reader);
+            });
+
+            services.AddSingleton(secureDatabase);
```
This seems odd to build the serviceProvider, modify stuff, and then continue adding to the services.  Is this okay to do?  I suspect a refactor could be done to make it cleaner

ViniciusSouza replied:
I refactored to get an instance from the factory, but also registered the type as I'm not sure that only the factory is in use.

mattklup commented:
Can you provide a high level description of the problem you are solving and how you plan to solve it?  The title doesn't give me enough context and security changes tend to be more complex than most

mattklup commented on src/Arriba/Arriba.Composition/Composition/Host.cs:
```diff
@@ -55,6 +55,14 @@ public static void AddArribaServices(this IServiceCollection services, ISecurity
+            var sp = services.BuildServiceProvider();
+            var databaseFactory = sp.GetService(typeof(DatabaseFactory)) as DatabaseFactory;
+
+            services.AddSingleton(databaseFactory.GetDatabase());
```
Why do we need both a factory and a singleton instance registered?  Whatever is getting the Database injected could use the factory?

ViniciusSouza replied:
We were registering both types, I kept doing the same but changed the source...

ViniciusSouza replied:
I tracked the usage of DatabaseSecure and was easy to remove it's dependencies, thanks ;)

ericmaino commented on src/Arriba/Arriba.Composition/Composition/Host.cs:
```diff
@@ -55,6 +56,26 @@ public static void AddArribaServices(this IServiceCollection services, ISecurity
 
+        private static void AddSecureDatabase(this IServiceCollection services)
+        {
+            var sp = services.BuildServiceProvider();
```
Using BuildServiceProvider() is a red flag to me. There are cases where this should be used, however I don't think this is one of them. I might suggest rewriting this as follows.
```suggestion
            services.AddSingleton<SecureDatabase>(p => {
                 var databaseFactory = p.GetService<DatabaseFactory>();

                var secureDatabase = databaseFactory.GetDatabase();

            //HACK: As the crawler don't have the AppRoles available at the JWT Token as claims
            //needed to grant for each role (PermissionScope) the roles name must use the same name as the Scope.
            secureDatabase.TableNames.ToList().ForEach(table =>
            {
                var security = secureDatabase.Security(table);
                security.Grant(IdentityScope.Group, PermissionScope.Owner.ToString(), PermissionScope.Owner);
                security.Grant(IdentityScope.Group, PermissionScope.Writer.ToString(), PermissionScope.Writer);
                security.Grant(IdentityScope.Group, PermissionScope.Reader.ToString(), PermissionScope.Reader);
            });

            return secureDatabase;
          });
```

ericmaino commented on src/Arriba/Arriba.Server/Extensions/OAuthExtension.cs:
```diff
@@ -2,22 +2,26 @@
+    {
         public static IServiceCollection AddOAuth(this IServiceCollection services, IArribaServerConfiguration serverConfig)
         {
+            services.AddSingleton<IAuthorizationHandler, AllowAnonymous>();
```
This line confused me. I think it has todo do with the filter being called AllowAnonymous. I would suggest renaming maybe to **ArribaConfigurationAuthorizationOverride**. This indicates is something in this project.

ViniciusSouza replied:
I did use ArribaConfigurationAuthorizationOverrideHandler at the handlers folder, also register it at Startup ConfigureService to be easy to find it

ericmaino commented on src/Arriba/Arriba.Server/Extensions/OAuthExtension.cs:
```diff
@@ -28,9 +32,28 @@ public static IServiceCollection AddOAuth(this IServiceCollection services, IArr
+        {
+            if (!_serverConfiguration.EnabledAuthentication)
+            {
+                foreach (IAuthorizationRequirement requirement in context.PendingRequirements.ToList())
```
Does this only work if it's the first or last handler?

ViniciusSouza replied:
[Filters in ASP.NET Core](https://docs.microsoft.com/en-us/aspnet/core/mvc/controllers/filters?view=aspnetcore-3.1 )
Authorization filters run first and are used to determine whether the user is authorized for the request. Authorization filters short-circuit the pipeline if the request is not authorized.

To use the [Authorize] attribute on the Controller/Endpoint this was the way I came up with. The way it was implemented before if we try to use [Authorize] attribute with the Authentication settings disabled we will get an exception for trying to use Authorization without the service being registered.

ericmaino replied:
The question is how does the filter run if there are multiple authorization filters installed. Does the order matter?

ViniciusSouza replied:
So Eric this is not a Filter it is a Handler and we can have many instances as we want to all of them are going to be executed. 

https://docs.microsoft.com/en-us/aspnet/core/security/authorization/policies?view=aspnetcore-3.1

```
public async Task<AuthorizationResult> AuthorizeAsync(ClaimsPrincipal user, 
             object resource, IEnumerable<IAuthorizationRequirement> requirements)
{
    // Create a tracking context from the authorization inputs.
    var authContext = _contextFactory.CreateContext(requirements, user, resource);

    // By default this returns an IEnumerable<IAuthorizationHandlers> from DI.
    var handlers = await _handlers.GetHandlersAsync(authContext);

    // Invoke all handlers.
    foreach (var handler in handlers)
    {
        await handler.HandleAsync(authContext);
    }

    // Check the context, by default success is when all requirements have been met.
    return _evaluator.Evaluate(authContext);
}
```

ericmaino commented on src/Arriba/Arriba.Server/Middlewares/ArribaNormalizeIdentityMiddleware.cs:
```diff
@@ -0,0 +1,60 @@
+                if (claim.Type == ClaimTypes.Role)
+                {
+                    addClaim = false;
+                    claims.Add(new Claim(ClaimTypes.Role, claim.Value.ToLowerInvariant()));
```
Why use ToLowerInvariant()

ViniciusSouza replied:
Because it will depend in how the crawler or the hack deals with the database security groups, it should me be Lower() as the Hack is doing.

ericmaino replied:
Shouldn't let that affect this. If the values are case sensitve then use a constant to replace the value. If case sensitivity does not matter then if this breaks, fix the bug elsewhere in the code to not be case dependent.

ViniciusSouza replied:
I changed the class SecurityIdentity, it was changing the Identity name to ToLowerInvariant() and now I'm using the name as received, also changed the tests that was impacted.

ericmaino commented on src/Arriba/Arriba.Server/Middlewares/ArribaNormalizeIdentityMiddleware.cs:
```diff
@@ -0,0 +1,60 @@
+            await _next(httpContext);
+        }
+
+        private ClaimsPrincipal NormalizeIdentity(ClaimsPrincipal user)
```
Seems like this function could be extracted into another class and then easily tested

ericmaino commented on src/Arriba/Arriba/Diagnostics/Tracing/ArribaLogs.cs:
```diff
@@ -16,5 +16,9 @@ public void WriteLine(string message, params object[] args)
             Console.WriteLine(message, args);
         }
 
+        public void WriteLine(string message)
```
Why add this?

ViniciusSouza replied:
After getting the changes merged I was getting an error  just during execution at this call on OAuthController.ReadTokenResultAsync

``` 
_log.WriteLine($"ReadTokenResultAsync result: {result}");
```

Instead of changing this call I decided to add a method that receives the string

ericmaino replied:
hmm.. weird to see it be required, but not to see a change that uses is..

ViniciusSouza replied:
I did change the usage of the logger and removed the method overload

ViniciusSouza commented:
> Can you provide a high level description of the problem you are solving and how you plan to solve it? The title doesn't give me enough context and security changes tend to be more complex than most

The Microsoft.AspNetCore.Authentication.JwtBearer library already make available all the JWT Token properties to ClaimsIdentity in form of claims. Some properties that Arriba is using for IPrincipal Identity was not available, name for instance and the AppRoles names, that will depend in how the crawler is adding them.

ViniciusSouza committed:
FIX: Use DatabaseFactory to get instance of SecureDatabase

ViniciusSouza committed:
REFACTOR: Remove ToLowerInvariant from SecurityIdentity name

ViniciusSouza committed:
FIX: RevokeAccess Unit test using GrantAccess methods

ViniciusSouza committed:
FIX: NGINX route for oauth controller

The OAuth route will be used for both versions V1 and V2

ViniciusSouza committed:
FIX: Remove singleton register for ISecurityConfiguration

ServerConfiguration already implement ISecurityConfiguration

ericmaino commented on src/Arriba/Arriba.Server/Controllers/OAuthController.cs:
```diff
@@ -94,7 +96,7 @@ private async Task<OAuthTokenResult> ReadTokenResultAsync(Uri uri, HttpContent c
             var result = await response.Content.ReadAsStringAsync();
 
-            _log.WriteLine($"ReadTokenResultAsync result: {result}");
+            _log.WriteLine("ReadTokenResultAsync result: {0}", result);
```
The inline string interpolation would be preferred. Instead of changing this seems that _log should have exposed a single WriteLine(string x)

ViniciusSouza replied:
I did that before.. by adding the WriteLine(string x), because I was getting a Error during the runtime (and just at runtime) it tries to use the params with 0 items on it. 

I thought that you wanted to change it to use what was available on _log.. I will redo the changes  ;)

ericmaino commented on .docker/WebUI/Configuration/Configuration.jsx:
```diff
@@ -2,11 +2,10 @@
+        // When using NGINX as Reverse Proxy it can be setup as the frontend url plus /v2 or v1 
+        url: window.location.protocol + "//" + window.location.host + "/v2",
+        
+        oauthUrl: window.location.protocol + "//" + window.location.host + "/oauth/",
```
I like this change, but really don't like the duplication

ViniciusSouza replied:
I tried to use a baseUrl property and reuse at the next line, but it didn't work. I will look to a different way to do that because it was bordering me as well

ViniciusSouza replied:
I had to create a variable before the export.

```
    let baseArribaUrl = window.location.protocol + "//" + window.location.host;
```

then at the export

```
url:  baseArribaUrl + "/v2",
        
oauthUrl: baseArribaUrl + "/oauth/",

```

ericmaino commented on src/Arriba/Arriba.Server/Extensions/OAuthExtension.cs:
```diff
@@ -28,9 +32,28 @@ public static IServiceCollection AddOAuth(this IServiceCollection services, IArr
                 auth.DefaultPolicy = jwtBearerPolicy;
             });
 
-            services.AddSingleton((_) => serverConfig.OAuthConfig);
```
Is this getting registered somewhere else or was it a duplicate registration? Seems like stuff would break with this change

ViniciusSouza replied:
It is registered on the startup

ericmaino commented on src/Arriba/Arriba/Server/Authorization/ArribaIdentityNormalizer.cs:
```diff
@@ -0,0 +1,36 @@
+        public ClaimsPrincipal NormalizeIdentity(ClaimsPrincipal user, bool throwExceptionWhenNameNotFound = true)
+        {
+            user.ThrowIfNull(nameof(user));
+            ParamChecker.ThrowIfNull(user.Identity, nameof(user.Identity));
```
seems like this line could be for consistency
```suggestion
            user.Identity.ThrowIfNull(nameof(user.Identity));
```

ViniciusSouza replied:
We don't have a extension for the Identity, but the ParamChecker has the thowIfNull using generics.
I was in doubt at the time to create an extension or use the generic one, at the end I preferred to use the generic because the others have a higher use frequency.

ViniciusSouza replied:
I added a guard method for IIdentity

ericmaino commented on src/Arriba/Arriba/Server/Authorization/ArribaIdentityNormalizer.cs:
```diff
@@ -0,0 +1,36 @@
+            ParamChecker.ThrowIfNull(user.Identity, nameof(user.Identity));
+            user.Identity.AuthenticationType.ThrowIfNullOrWhiteSpaced(nameof(user.Identity.AuthenticationType));
+
+            var nameClaim = user.Claims.Where(c => c.Type == ClaimTypeName).FirstOrDefault();
```
Should use String.Equals, likely with case insensitivity. Do does it matter if this is case sensitive?

ViniciusSouza replied:
I Identity.HasClaims is case sensitive and the database security too.

ericmaino commented on src/Arriba/Arriba/Server/Authorization/IArribaIdentityNormalizer.cs:
```diff
@@ -0,0 +1,14 @@
+        /// </summary>
+        /// <param name="identity"></param>
+        /// <returns></returns>
+        ClaimsPrincipal NormalizeIdentity(ClaimsPrincipal identity, bool throwExceptionWhenNameNotFound);
```
throwExceptionWhenNameNotFound seems like this is always true in code, besides tests. If that is the case the it could likely be removed.

ViniciusSouza replied:
At the time I was thinking to give an option to not throw the exception.

ViniciusSouza committed:
FIX: Adding WriteLine(string str) on ArribaLog

ericmaino commented on src/Arriba/Arriba/ParametersCheckers/ParamChecker.cs:
```diff
@@ -53,5 +53,11 @@ public static void ThrowIfNullOrEmpty(this NameValueCollection value, string par
                 throw new ArgumentException("Not Provided", paramName);
         }
+
+        public static void ThrowIfNull(this IIdentity value, string paramName)
```
This should not be required. Actually the param check above in lines 11-22 should be consolidated. The only one that should need to exist is the ThrowIfNull<T>(this T value, string name).

The conditional check should also be: if (value is null)

ViniciusSouza replied:
To use the is null for the T I do have to use the version 8.0 for the language definition on Arriba, is that okay?

ViniciusSouza committed:
REFACTOR: Consolidate null check

ViniciusSouza committed:
FEATURE: Normalize identity's claims for Arriba Authorization

It adds a middleware that calls IArribaIdentityNormalizer.NormalizeIdentity
performing any change it may be necessary the current implementation
of this interface creates a new Identity using the JWT Claims Name
that uses a ClaimsType Name.

ViniciusSouza committed:
FIX: Missing this for extension use is to check null
